### ISSUE TITLE
关于mnist数据集和train_mnist.py的一些分析

### ISSUE THREAD

**Issue by leewishyuanfang** [huaweicloud/ModelArts-Lab#5]
mnist数据集：
手写体数字识别数据集,是NIST数据集的一个子集,包含了60000张图片作为训练数据,10000张图片作为测试数据。其中，每一张图片都代表了0-9中的一个数字，图片的大小都为28*28，且数字都会出现在图片的正中间。

train-images-idx3-ubyte.gz：训练数据图片文件
train-labels-idx1-ubyte.gz：训练数据标签文件（即每一张图片对应的具体数字）
t10k-images-idx3-ubyte.gz：测试数据图片
t10k-labels-idx1-ubyte.gz：测试数据标签文件（即每一张图片对应的具体数字）

**Comment by leewishyuanfang** [comment]
train_mnist.py分析：
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

 # input_data是TensorFlow提供的专门用来处理MNIST数据的类，这个类会自动下载和转化MNIST数据的格式，将数据从原始的数据包中解析成训练和测试神经网络时使用的格式。
from tensorflow.examples.tutorials.mnist import input_data
# 导入tensorflow
import tensorflow as tf
# 这个应该是作者自定义的模块，已经部署在华为云上了
import moxing.tensorflow as mox
import os

tf.flags.DEFINE_string('data_url', None, 'Dir of dataset')
tf.flags.DEFINE_string('train_url', None, 'Train Url')

flags = tf.flags.FLAGS


def check_dataset():
  work_directory = flags.data_url
  # mnist数据集中的四个文件名
  filenames = ['train-images-idx3-ubyte.gz', 'train-labels-idx1-ubyte.gz', 't10k-images-idx3-ubyte.gz',
               't10k-labels-idx1-ubyte.gz']
  # 检查在指定目录中，mnist数据集中的四个文件是否存在
  for filename in filenames:
    filepath = os.path.join(work_directory, filename)
    if not mox.file.exists(filepath):
      raise ValueError('MNIST dataset file %s not found in %s' % (filepath, work_directory))


def main(*args, **kwargs):
  # 检测数据集的有效性
  check_dataset()
  # 导入数据集文件，并将其转化为mnist对象，这个对象的格式是适合训练和测试网络的格式
  # 处理后的MNIST数据集划分为train、validation和test三个数据集。处理后的每一张图片是一个长度为784的一维数组，这个数组对应了图片像素矩阵中的每一个数字（28*28=784）。像素矩阵中元素的取值范围为[0,1]，它代表了颜色深浅（0为白色，1为黑色）。
  # 通过mnist对象，可以分别获得train、validation和test这三个数据集
  mnist = input_data.read_data_sets(flags.data_url, one_hot=True)

  # 定义输入数据集，返回图像集和标签集
  # define the input dataset, return image and label
  def input_fn(run_mode, **kwargs):
    def gen():
      while True:
        # mnist.train.next_batch函数，它可以从所有的训练数据中读取一小部分（程序中是50）作为一个训练batch
        yield mnist.train.next_batch(50)
    ds = tf.data.Dataset.from_generator(
        gen, output_types=(tf.float32, tf.int64),
        output_shapes=(tf.TensorShape([None, 784]), tf.TensorShape([None, 10])))
    return ds.make_one_shot_iterator().get_next()

  # 定义模型，本次的模型是简单的、一层的全连接神经网络，也就是所谓的softmax_regression
  # define the model for training or evaling.
  def model_fn(inputs, run_mode, **kwargs):
    # inputs就是input_fn函数返回的图像和标签（一个batch），y_即理想的输出
    x, y_ = inputs
    # 全连接神经网络的权重，初始化为全0
    W = tf.get_variable(name='W', initializer=tf.zeros([784, 10]))
    # 全连接神经网络的偏置，也初始化为全0
    b = tf.get_variable(name='b', initializer=tf.zeros([10]))
    # 把输入x与权重进行矩阵相乘，再加上偏置，就得到了神经网络的实际输出y，这也就是前向传播过程
    # 输入x是一个长度为784的一维数组，代表一张手写数字图像，
    y = tf.matmul(x, W) + b
    # 根据实际输出y和理想输出y_计算损失函数，理论上希望损失函数的值越小越好，本程序使用的是softmax损失函数
    cross_entropy = tf.reduce_mean(
      tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y))
    # 计算模型在当前数据集batch上的正确率
    correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
    # 通过正确率的均值计算得到模型的准确度
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
    # 此处代码涉及作者自己的moxing模块，但意思应该是在此处给输入和输出进行实际的赋值，之前程序中的x,y只是一个placeholder，也就是占位符
    export_spec = mox.ExportSpec(inputs_dict={'images': x}, outputs_dict={'logits': y}, version='model')
    # 此处不是很明白，主要是缺少对moxing模块的了解
    return mox.ModelSpec(loss=cross_entropy, log_info={'loss': cross_entropy, 'accuracy': accuracy},
                         export_spec=export_spec)

  # 执行训练过程
  mox.run(input_fn=input_fn,                                                                  # 输入数据
          model_fn=model_fn,                                                                    # 训练模型
          optimizer_fn=mox.get_optimizer_fn('sgd', learning_rate=0.01),  # 指定反向传播（模型优化）算法和学习率
          run_mode=mox.ModeKeys.TRAIN,                                               # 表明当前是进行模型训练
          batch_size=50,                                                                              # 每次训练使用的数据集batch的大小（即一次训练用到的图像数量）
          auto_batch=False,
          log_dir=flags.train_url,                                                                  # 训练模型和日志的存放位置
          max_number_of_steps=1000,                                                       # 最大训练次数 
          log_every_n_steps=10,                                                                  # 每训练10次保存一次模型和日志
          export_model=mox.ExportKeys.TF_SERVING)                               # 导出模型的格式？
# 程序执行的入口
if __name__ == '__main__':                                     
    tf.app.run(main=main)

**Comment by chenliang613** [comment]
Great, thanks for your good sharing!

**Comment by chenliang613** [comment]
@leewishyuanfang  期待你的实践点2输出 ：）

实践点2：按照下面make_your_mnist.py代码，修改代码中的文件路径，在本地制作适配mnist的图片数据，来测试模型的精准度。（其中使用的原图需要黑底白字）
https://github.com/huaweicloud/ModelArts-Lab/blob/master/%E7%BB%BC%E5%90%88AI%E5%BC%80%E5%8F%91%E6%A1%88%E4%BE%8B/MNIST%20Dataset%20Recognition/mnist_pic/make_your_mnist.py
